### ISSUE TITLE
Enhancement Request - More delay options

### ISSUE THREAD

**Issue by lincolnVaino** [Necrobot-Private/NecroBot#732]
To make bot more undetectable by Niantic, bot should have more detailed delays. Currently you can set delay for action, transfer, upgrade, catch, recycle, rename and evolve. This does not however correlate correctly with human beahviour and for that reason, makes bot detectable.

There should be own delays for pokestop spin (currently ties with action delay), delay before capture (currently ties with action delay), delay after login (currently ties with action delay) and delay for start of movement (random pauses at pokestops?).

**Consider this scenario (how to act like a human vs how delays are handled):**
1. Login
2. Select pokestop and spin (Action delay)
3. Engage pokemon (Action delay)
4. Throw pokeballs (Action delay)
5. Wait for capture animation (not including the capture itself) (Delay between pokemon capture)
6. Pokemon caught (Action delay)
7. Spin new stop (Action delay)
8. Start moving (No delays currently)

**Example of real life scenario:**
1. Login, 15 seconds till loaded and see pokestops and pokemons
2. Spin pokestop 1.5 seconds
3. Engage pokemon 1 seconds
4. Throw and wait for capture animation (fail or succeed)  6 seconds
5. Pokemon caught 6 seconds
7. Spin pokestop 1.5 seconds
8. Start moving

**As stated above, if you have action delay 1.5 seconds (as it is time to spin pokestop) then bot acts like this:**
Login, 1.5s till spin (Game would still be loading), 1.5s engage and capture animation (This is impossible within client), 6s delay between catch (Acts like normal client)
spin 1.5s (Acts like normal client in this case after the capture).

**So if we could set more delays for variety of things stated at start of this issue, route would be like this:**
1. Login 15s (AfterLoginDelay => Acts like normal client)
2. Spin 1.5 (SpinDelay => Acts like normal client)
3. Engage and attempt capture (EngageCatchDelay => Acts like normal client)
4. Caught message (DelayBetweenPokemonCatch => Acts like normal client, this is already implemented)
5. Spin 1.5s (SpinDelay)
6. Move (DelayBetweenMovement)

**If you want to act like normal client, you have to set DelayBetweenPlayerActions to atleast 6 seconds to seem human at all. Under that, you are doing things that are not possible within real client.**

**Adittion to this is bulk transfer delay:** currently bot can select 250 pokemons within 2 seconds:
1. Select 100 pokemons for bulk transfer, 1s (Plausible if there was no spins or captures)
2. Select 100 pokemons for bulk transfer, 1s (Detectable, no human can ever select 100 pokemons within client inside 1 second timeframe)
3. Select 50 pokemons for bulk transfer, 1s (Detectable, no human can ever select 50 pokemons within client inside 1 second timeframe)

**So tl;dr;**
Make these delay types available (basicly detach to own delays from Action Delay):
1. SpinDelay
2. EngageCatchDelay
3. AfterLoginDelay
4. DelayBetweenMovement
5. BulkTransferDelayBetweenBuffer

**Edit:** I can draw detailed pictures if needed or try to explain better :)

**How to try yourself:** Play and stopwatch every step you do them in real client (Pokemon GO) then try to 
configure bot delays to correlate with your stopwatch times and you realize there is not enough options.

**Comment by bulubebek** [comment]
+1 for this. :)

**Comment by jjskuld** [comment]
If you give me a list of the various timings from the client to then I'll implement it.

**Comment by lincolnVaino** [comment]
These all should be user configurable :)

**BeforeCatchDelay**: 1500ms - 2500ms
**CatchMissedDelay**: 1000ms
**CatchEscapeDelay**: 3500ms - 6000ms
**CatchSuccessDelay**: 13000ms
**CatchFleeDelay**: 2000ms
**CatchSuccessDelay**: 130000ms

**AfterLoginDelay**: 6000ms - 50 000ms (This varies based on users phone. Iphone 7 takes six seconds, old android phone takes 50 seconds...)

**PokeStopDelay**: 1500ms-2000ms

**BeforeBulkTransferDelay**: (number of pokemons to transfer) * 200ms
  Example: select 10 pokemons and transfer: 2000ms
                 select 100 pokemons and transfer: 20 000ms

**Image to clarify where delays should be:**
![drawgo](https://cloud.githubusercontent.com/assets/25554500/22651350/765eb8b2-ec8b-11e6-8952-54ec5c1936e5.PNG)

**Comment by jjskuld** [comment]
lincolnVaino Thanks for doing this.. I'll see if I can implement it tonight.

**Comment by RoskamiesX** [comment]
Implemented: https://github.com/Necrobot-Private/NecroBot/pull/1026

**Comment by bulubebek** [comment]
great! thank you mate.

**Comment by RoskamiesX** [comment]
You can now use ActionDelay for pokestop delays, no login delays, but i think that it is actually not needed. Never received ban when using high delay configs. So i'd say that these issue can be closed 👍

**Comment by hardknoxni69a** [comment]
implemented
